File: engines/rendering/blink/src/page.cpp
/*
 * Nexus Browser - Blink Rendering Engine
 * Copyright (C) 2026
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#include "page.hpp"

#include <algorithm>

#include "frame.hpp"
#include "rendering_context.hpp"

namespace nexus::blink {

std::uint64_t Page::next_page_id_ = 1;

Page::Page(const std::shared_ptr<RenderingContext>& context)
    : page_id_(next_page_id_++)
    , rendering_context_(context)
{
    initialize();
}

Page::~Page()
{
    close();
}

bool Page::initialize()
{
    if (open_) {
        return true;
    }

    if (!rendering_context_) {
        return false;
    }

    if (!rendering_context_->is_initialized()) {
        if (!rendering_context_->initialize()) {
            return false;
        }
    }

    main_frame_ = std::make_shared<Frame>(rendering_context_);

    if (!main_frame_->initialize()) {
        main_frame_.reset();
        return false;
    }

    open_ = true;

    return true;
}

bool Page::close()
{
    if (!open_) {
        return true;
    }

    for (auto& frame : child_frames_) {
        if (frame) {
            frame->close();
        }
    }

    child_frames_.clear();

    if (main_frame_) {
        main_frame_->close();
        main_frame_.reset();
    }

    open_ = false;

    return true;
}

bool Page::navigate(const std::string& url)
{
    if (!open_ || !main_frame_) {
        return false;
    }

    return main_frame_->navigate(url);
}

bool Page::reload()
{
    if (!open_ || !main_frame_) {
        return false;
    }

    return main_frame_->reload();
}

std::shared_ptr<Frame> Page::main_frame() const
{
    return main_frame_;
}

bool Page::add_frame(const std::shared_ptr<Frame>& frame)
{
    if (!open_ || !frame) {
        return false;
    }

    if (!frame->is_active()) {
        if (!frame->initialize()) {
            return false;
        }
    }

    child_frames_.push_back(frame);

    return true;
}

bool Page::remove_frame(const std::shared_ptr<Frame>& frame)
{
    if (!frame) {
        return false;
    }

    auto iterator = std::find(
        child_frames_.begin(),
        child_frames_.end(),
        frame
    );

    if (iterator == child_frames_.end()) {
        return false;
    }

    (*iterator)->close();
    child_frames_.erase(iterator);

    return true;
}

std::size_t Page::frame_count() const
{
    if (!open_) {
        return 0;
    }

    return 1 + child_frames_.size();
}

std::uint64_t Page::page_id() const
{
    return page_id_;
}

bool Page::is_open() const
{
    return open_;
}

} // namespace nexus::blink

File: engines/rendering/servo/src/page.rs
/*
 * Nexus Browser - Servo Page
 * Copyright (C) 2026
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

use crate::frame::Frame;

pub struct Page {
    id: u64,
    main_frame: Frame,
    child_frames: Vec<Frame>,
    open: bool,
}

impl Page {
    pub fn new(
        page_id: u64,
        frame_id: u64,
    ) -> Self {
        let mut main_frame = Frame::new(frame_id);
        main_frame.initialize();

        Self {
            id: page_id,
            main_frame,
            child_frames: Vec::new(),
            open: true,
        }
    }

    pub fn close(&mut self) {
        for frame in &mut self.child_frames {
            frame.close();
        }

        self.child_frames.clear();

        self.main_frame.close();

        self.open = false;
    }

    pub fn navigate(
        &mut self,
        url: &str,
    ) -> Result<(), String> {
        if !self.open {
            return Err(
                "Page is closed".to_string()
            );
        }

        self.main_frame.navigate(url)
    }

    pub fn reload(&mut self) -> Result<(), String> {
        if !self.open {
            return Err(
                "Page is closed".to_string()
            );
        }

        self.main_frame.reload()
    }

    pub fn add_frame(
        &mut self,
        mut frame: Frame,
    ) -> Result<(), String> {
        if !self.open {
            return Err(
                "Page is closed".to_string()
            );
        }

        frame.initialize();

        self.child_frames.push(frame);

        Ok(())
    }

    pub fn id(&self) -> u64 {
        self.id
    }

    pub fn main_frame(&self) -> &Frame {
        &self.main_frame
    }

    pub fn main_frame_mut(&mut self) -> &mut Frame {
        &mut self.main_frame
    }

    pub fn frame_count(&self) -> usize {
        if !self.open {
            return 0;
        }

        1 + self.child_frames.len()
    }

    pub fn is_open(&self) -> bool {
        self.open
    }
}
